### ISSUE TITLE
When executing phpstan for a given file, include the used traits as well

### ISSUE THREAD

**Issue by BladeMF** [phpactor/phpactor#2717]
Otherwise it gives false positives about unused private methods which are declared as abstract in the trait.

**Comment by dantleech** [comment]
I would guess this only happens if the traits have nkt been persisted on the filesystem..

**Comment by BladeMF** [comment]
What do you mean? The scenario is:
1. I am in a class that uses a trait (the trait exists)
2. I have a private method that is required by the trait
3. phpstan reports an error for those methods as unused, because it is only checking the file, not accounting for the existence of the trait which actually uses those methods. That's why when running phpstan on the whole project, no such error is reported.

**Comment by dantleech** [comment]
It should take kntk account persisted files, I.e. the traits.

**Comment by BladeMF** [comment]
Don't we copy the file into tmp and run it there? I have a fuzzy memory of seeing that.

**Comment by dantleech** [comment]
Yes the version in your editor is copied to a. Tmp.directory, but phpstan is still aware of all the other files in the project

**Comment by BladeMF** [comment]
Mmmm, are you positive about that? I can take it up with them then.

**Comment by dantleech** [comment]
> because it is only checking the file, not accounting for the existence of the trait which actually uses those methods

if it were only taking into account the code in a single file, then the PHPStan would complain each time you extend a class or use a trait.

Can you provide some code that reproduces it.

**Comment by BladeMF** [comment]
Sure, I'll try.
